src/l1_fetcher.rs

```diff
@@ -184,6 +182,22 @@ impl L1Fetcher {
 
         Ok(())
     }
+
+    async fn get_logs(&self, filter: &Filter) -> Result<Vec<Log>> {
+        for attempt in 1..MAX_RETRIES + 1 {
+            match self.provider.get_logs(&filter).await {
+                Ok(x) => return Ok(x),
+
+                Err(e) => {
+                    tracing::error!("attempt {attempt}: failed to get logs: {e}");
+
+                    sleep(Duration::from_millis(50 + random::<u64>() % 500)).await;
+                }
+            }
+        }
+
+        Err(L1GetLogsError {}.into())
+    }
```

Comment:
```suggestion
    async fn retry_call<T>(&self, callback: Fn) -> Result<Vec<T>> {
        for attempt in 1..MAX_RETRIES + 1 {
            match callback {
                Ok(x) => return Ok(x),
                Err(e) => {
                    tracing::error!("attempt {attempt}: failed to get value from L1: {e}");
                    sleep(Duration::from_millis(50 + random::<u64>() % 500)).await;
                }
            }
        }
    }
```
We can make this more generic, by passing in a callback function instead and matching on that.

Reply:
What about the log message + error type after failed iterations?

Reply:
Probably make a parent Enum, something like `enum L1FetchTimeoutError`. Then we just return that type instead

Reply:
Another thing: Plain `Fn` doesn't work. How do you define the same but as `async fn` ?

Reply:
You can use this syntax:

```rust
async fn some_function(callback: impl Fn() -> Fut)
where
    Fut: Future<Output = Result<T>>,
}
```

Then it's just `await`ing the closure inside the function as always. However, this does add a lot of complexity so if you don't anticipate that we would need any more, retry fetches from L1, we can just defer this problem for another time.

Reply:
One last problem: How to handle the issue with differing error types?

> this does add a lot of complexity so if you don't anticipate that we would need any more, retry fetches from L1, we can just defer this problem for another time

Honestly, I don't think anything justifies this since the generic version is longer, a lot more complex and more difficult to read, compared to original in-place one, but now that we are on this road, I want to see this through :laughing:

Reply:
Let's do this then! :handshake:

Reply:
> One last problem: How to handle the issue with differing error types?
If we implement those errors under an `enum` that won't be an issue, or am I misunderstanding something?